arlo_server/rounds.py

```diff
@@ -0,0 +1,155 @@
+    if r["roundNum"] != next_round_num:
+        raise BadRequest(f"The next round should be round number {next_round_num}")
+
+    if r["roundNum"] == 1 and not r["sampleSize"]:
+        raise BadRequest(f"Sample size is required for round 1")
+
+
+def sample_ballots(election: Election, round: Round, sample_size: int = None):
+    # For now, we only support one targeted contest
+    contest = next(c for c in election.contests if c.is_targeted)
+
+    # Compute the total number of ballot samples in all rounds leading up to
+    # this one. Note that this corresponds to the number of SampledBallotDraws,
+    # not SampledBallots.
+    previous_rounds = (
+        db.session.query(RoundContest)
+        .join(Round)
+        .filter(Round.election_id == election.id, Round.id != round.id)
+        .all()
+    )
+    num_sampled = sum(r.sample_size for r in previous_rounds)
+
+    # Create the pool of ballots to sample (aka manifest) by combining the
+    # manifests from every jurisdiction in the contest's universe.
+    # Audits must be deterministic and repeatable for the same real world
+    # inputs. So the sampler expects the same input for the same real world
+    # data. Thus, we use the jurisdiction and batch names (deterministic real
+    # world ids) instead of the jurisdiction and batch ids (non-deterministic
+    # uuids that we generate for each audit).
+    manifest = {
+        (jurisdiction.name, batch.name): batch.num_ballots
+        for jurisdiction in contest.jurisdictions
+        for batch in jurisdiction.batches
+    }
+    batch_key_to_id = {
+        (jurisdiction.name, batch.name): batch.id
+        for jurisdiction in contest.jurisdictions
+        for batch in jurisdiction.batches
+    }
+
+    # Do the math! I.e. compute the actual sample
+    sample = sampler.draw_sample(
+        election.random_seed, manifest, sample_size, num_sampled
+    )
+
+    # Record which ballots are sampled in the db
+    for (ticket_number, (batch_key, ballot_position), times_sampled) in sample:
+        batch_id = batch_key_to_id[batch_key]
+        if times_sampled == 1:
+            sampled_ballot = SampledBallot(
+                batch_id=batch_id, ballot_position=ballot_position,
+            )
+            db.session.add(sampled_ballot)
+
+        sampled_ballot_draw = SampledBallotDraw(
+            batch_id=batch_id,
+            ballot_position=ballot_position,
+            round_id=round.id,
+            ticket_number=ticket_number,
+        )
+        db.session.add(sampled_ballot_draw)
+
+    db.session.commit()
+
+
+@app.route("/election/<election_id>/round", methods=["GET"])
+def list_rounds(election_id: str):
+    election = Election.query.get_or_404(election_id)
+    require_audit_admin_for_organization(election.organization_id)
+
+    rounds = (
+        Round.query.filter_by(election_id=election_id).order_by(Round.round_num).all()
+    )
+
+    return jsonify({"rounds": [serialize_round(r) for r in rounds]})
```

Comment:
@eventualbuddha you convinced me to have a top-level wrapper object. I know I convinced you not to, lol, but I think you're right that we should move in this direction

Reply:
For an array I really think it is necessary, and then it kind of trickles out…

Reply:
Do you also think it's a good practice for the body of POST/PUT requests? Or can those take the data without any wrapper?

Reply:
+1 on wrapping arrays in an object. It stems from an old security bug with cross-site JavaScript, which I believe is mostly fixed, but I think is worthwhile sticking to :)